Add prterun manpage, cleanup other manpages

jjhursey opened the pull request:
* Add a `prterun.1` manpage that mostly references other pages to
   avoid too much redundancy.
 * Split out the map/rank/bind discussion to a stand alone man
   page since it is shared by `prun` and `prterun` (and, one could
   argue `PMIx_Spawn` in the context of PRTE). Additionally, it
   is quite expansive so this provides a single place to provide
   additional examples and notes on the interplay of these three
   operations.
   - Add a soft link for `prte-rank.1` and `prte-bind.1` which
     link to `prte-map.1` for ease of use.
 * Closes #450

jjhursey commented:
This PR reflects the behavior of the master branch as of today. It does not reflect the changes that we wish to make as noted [here](https://github.com/openpmix/prrte/issues/709#issuecomment-760489283). We can update the man pages as those changes come in.

jjhursey commented:
There are three open issues noted in the documentation:
 * https://github.com/openpmix/prrte/issues/770
 * https://github.com/openpmix/prrte/issues/771
 * https://github.com/openpmix/prrte/issues/772

I can remove those examples from the man pages for now, and we can add them back when the tickets are fixed. Or leave the examples, and when those tickets are fixed they can remove the `TODO` comment. I'd prefer the latter - that way the author can update the example as needed to reflect actual behavior.

jjhursey commented:
This should be ready for review now (had to fix a couple of build issues). It's probably easier to view the whole files than try to read the diffs since a bunch of stuff moved around.

jjhursey commented:
PR #774 added the qualifier `DONOTRESOLVE` to `--map-by`

Note to self - there may be other qualifiers that need to be documented. I just picked up the ones advertised in the `--help` output.

jjhursey commented:
Looking at the man page after my comment above I noticed a few other deprecated options that needed to be listed. I also cleaned up the rankfile examples.

I commented out the examples associated with open bugs. That makes the manpage correct when rendered, and easy for us to add back those examples once the bugs are fixed.

drwootton commented on src/tools/prte/prte.1.md:
```diff
@@ -60,30 +95,31 @@ cluster to use for the DVM.
 
 :   Send arguments to various MCA modules. See the "MCA" section,
     below.
 
```
There's a --pmixmcam option below this that mentions a MCA parameter file set list. Is the 'arg0' supposed to be a path to the file? I'm not sure what it is. Also, maybe the same comments I made about global PMIx MCA parameters and contexts apply here. I'm also not sure I understand why I would have global vs context-specific settings here or how I would control that since I don't see any way to specify context.

jjhursey replied:
Good point. I'll clarify.
 * The `arg0` is supposed to be a file path
 * The global MCA's apply to all app contexts in a MIMD launch (`prterun -gpmixmca foo bar ./a.out : ./b.out`) whereas the non-global MCA's (for the most part) apply only to that specific app context (`prterun -pmixmca foo bar ./a.out : -pmixmca foo baz ./b.out`)

drwootton commented on src/tools/prun/prun.1.md:
```diff
@@ -443,34 +281,39 @@ The parser for the `-x` option is not very sophisticated; it does not
-:   Pass global MCA parameters that are applicable to all contexts.
+`--gpmixmca <key> <value>`
+
+:   Pass global PMIx MCA parameters that are applicable to all contexts.
```
What is the difference between global PMIx MCA parameters and what's passed with --pmixmca, partly because I'm unfamiliar with the context concept. Well, reading forward thru the page I see discussion of app contexts primarily for MPMD, so I think clarifies global (all contexts) vs settings for a specific app context, where I'm guessing the specific options would be in a context file

jjhursey replied:
I clarified 'contexts' here a bit to help:
 > :   Pass global PMIx MCA parameters that are applicable to all application contexts. `<key>` is the parameter name; `<value>` is the parameter value.

drwootton commented on src/tools/prun/prun.1.md:
```diff
@@ -443,34 +281,39 @@ The parser for the `-x` option is not very sophisticated; it does not
+`--pmixmca <key> <value>`
 
-:   Aggregate MCA parameter set file list.
+:   Send arguments to various PMIx MCA modules. See the "MCA" section,
```
Is this supposed to be the pathname to a file?

jjhursey replied:
Yep. I added a clarification.
```
:   Aggregate PMIx MCA parameter set file list. The `arg0` argument is a
    comma-separated list of tuning files. Each file containing MCA parameter
    sets for this application context.
```

rhc54 commented on src/tools/prte/prte-map.1.md:
```diff
@@ -0,0 +1,855 @@
+ - `INHERIT`
+ - `NOINHERIT` means `!INHERIT`
+ - `PE-LIST=a,b` comma-delimited ranges of cpus to use for this job processed as an unordered pool of CPUs
+ - `FILE=%s` (path to file containing sequential or rankfile entries).
```
Ah, wait - yes it does. Rankfile is now `--map-by rankfile:file=myfile`

jjhursey replied:
I can confirm that the current accepted syntax is: `--map-by rankfile:FILE=myrankfile`

jjhursey commented:
I just pushed a couple of additions
 * Added back the package relative rankfile example now that Issue #772 is fixed
 * Added the text from https://github.com/openpmix/prrte/issues/673#issuecomment-784572798

I'll work on the review comments this evening.

jjhursey commented:
Added back the MIMD example now that Issue #771 is fixed

jjhursey commented on src/tools/prun/prun.1.md:
```diff
@@ -44,22 +47,20 @@ want to use a command line of the following form:
+in a round-robin fashion by CPU slot. If running under a supported resource
+manager a hostfile is usually not required unless the caller wishes to further
+restrict the set of resources used for that job.
 
```
I took out the description of binding from `prun.1` and replaced it with a reference to `prte-map.1` so we don't have the same information in multiple places:
> Please note that PRTE automatically binds processes. See prte-map(1) for defaults for the mapping, ranking, and binding of processes.

jjhursey commented:
Interesting CI failed in cycle with [gist](https://gist.github.com/ibm-ompi/e0963c25cc1d80864e8bc0afc14a1ca4)
```
20:30:29 [cffe4b8c5458:03719] UNEXPECTED MESSAGE tag = 39680 from source prte-3cd6154b28f4-346@0:9
20:30:29 --------------------------------------------------------------------------
20:30:29 A received msg header indicates a size that is too large:
20:30:29 
20:30:29   Requested size:  909194548
20:30:29   Size limit:      16777216
20:30:29 
20:30:29 If you believe this msg is legitimate, please increase the
20:30:29 max msg size via the ptl_base_max_msg_size parameter.
20:30:29 --------------------------------------------------------------------------
20:30:29 
```

There are no code changes in this PR. But this is something we should track down.

bot:ibm:gnu:retest

jjhursey commented:
bot:ibm:gnu:retest

rhc54 commented:
Here is my take on that error, though I haven't investigated enough to prove/disprove it. You are running a lot of `prun` calls in tight sequence. The socket timeout is fairly long in comparison, and every so often the new `prun` gets an old socket that hasn't been harvested yet. So there is leftover cruft sitting in that socket which manifests itself as an "unexpected message" as soon as `prun` opens it. Because it is incomplete, the bytes are totally out of sequence from what we expect, resulting in our parsing it into a ridiculous size.

Issue would be: what happens if the leftover bytes _didn't_ result in a ridiculous size, and we actually tried to parse that as a legitimate message? My guess would be: ouch!

jjhursey commented:
I just pushed the clarification from https://github.com/openpmix/prrte/issues/673#issuecomment-786691918

jjhursey commented:
For the error I mentioned above I filed https://github.com/openpmix/prrte/issues/787 for further discussion/debug - since it's not related to this PR.

jjhursey commented:
**Reviewers**: This is ready for review.

There is one outstanding bug (Issue #770) with `max_slots` that I'm working on. I'll update the man page with that fix.

There are probably other improvements to the man page that we can bring in over time as well.

acolinisi commented on src/tools/prun/prun.1.md:
```diff
@@ -412,25 +238,25 @@ To manage files and runtime environment:
+
+:   Synonym for `--wdir`.
+
 `-x <env>`
```
Sounds like CLI details are in flux and tricky with OMPI/PRRTE separation (#795), but a simple "requires `--personality ompi`" in the description for `-x` here might save a lot of pain.

jjhursey replied:
Interesting. @rhc54 is `-x` intended to only be for the OMPI personality? It seems like something that should be part of the core.

jjhursey replied:
I now see the discussion [here](https://github.com/openpmix/prrte/pull/795#issuecomment-790953094) and filed Issue #801 to fix the code.

jjhursey commented:
We are starting to get conflicts with other PRs making changes to the man pages. Can I get a review on this so we can all work from the same base?

rhc54 reviewed (approved):
LGTM - we can always update as we go, but this is a solid base. Thx!
